### PR TITLE
C++: Split generated code in header and implementation files

### PR DESCRIPTION
**Pull request by tronical**
Default to generating a .cpp file now that holds the definitions of our generated code. This way those definitions (bindings code, etc.) can be compiled in parallel to other application code that includes the generated header file.

Furthermore, the new argument can be used by application developers to scale up and generate more .cpp files to parallel builds, if the output of the slint-compiler becomes larger.

### CODE REVIEW

File: internal/compiler/generator/cpp.rs [hunk 1]
```diff
@@ -82,17 +85,96 @@ mod cpp_ast {
...
+            let mut cpp_files = Vec::with_capacity(count);
+            if count == 0 {
+                return cpp_files;
+            }
+
+            let mut definitions = Vec::new();
+
+            let mut i = 0;
+            while i < self.definitions.len() {
+                if matches!(&self.definitions[i], Declaration::Function(fun) if fun.template_parameters.is_some())
```
Comment (ogoffart) [inline internal/compiler/generator/cpp.rs:108]: The type definition can also not be moved.

File: internal/compiler/generator/cpp.rs [hunk 2]
```diff
@@ -82,17 +85,96 @@ mod cpp_ast {
...
+                    continue;
+                }
+                definitions.push(self.definitions.remove(i));
+            }
+
+            // Any definition in the header file is inline.
+            self.definitions.iter_mut().for_each(|def| match def {
+                Declaration::Function(f) => f.is_inline = true,
+                Declaration::Var(v) => v.is_inline = true,
+                _ => {}
```
Comment (ogoffart) [inline internal/compiler/generator/cpp.rs:120]: what if count == 0?  we returned earlier.

File: internal/compiler/generator/cpp.rs [hunk 3]
```diff
@@ -82,17 +85,96 @@ mod cpp_ast {
     ///A full C++ file
     #[derive(Default, Debug)]
     pub struct File {
+        pub is_cpp_file: bool,
         pub includes: Vec<String>,
         pub after_includes: String,
         pub namespace: Option<String>,
         pub declarations: Vec<Declaration>,
+        pub resources: Vec<Declaration>, // For now inline, maybe #embed in the future
```
Comment (ogoffart) [inline internal/compiler/generator/cpp.rs:93]:
```suggestion
        pub resources: Vec<Declaration>,
```

I think the comment is confusing. they should not be `inline` if we split them in a translation unit.
They should be forward declared.

File: internal/compiler/generator/cpp.rs (hunk 2 above)
Comment (tronical) [reply internal/compiler/generator/cpp.rs:120]: Good catch - thanks.

File: internal/compiler/generator/cpp.rs (hunk 3 above)
Comment (tronical) [reply internal/compiler/generator/cpp.rs:93]: Okay :)

File: api/cpp/cmake/SlintMacro.cmake [hunk 4]
```diff
@@ -48,8 +57,18 @@ function(SLINT_TARGET_SOURCES target)
         set(scale_factor_target_prop "$<TARGET_PROPERTY:${target},SLINT_SCALE_FACTOR>")
         set(scale_factor_arg "$<IF:$<STREQUAL:${scale_factor_target_prop},>,,--scale-factor=${scale_factor_target_prop}>")
 
+        if (compilation_units GREATER 0)
+            list(APPEND cpp_files "${CMAKE_CURRENT_BINARY_DIR}/${_SLINT_BASE_NAME}.cpp")
```
[the remark this answers is not in the document]
Comment (tronical) [reply api/cpp/cmake/SlintMacro.cmake:61]: Good catch. We do this with the memory game, where it breaks even for out of source builds (memory.cpp is the "main.cpp" - a bit odd next to `main.js` and `main.rs`, but nevertheless a bug here).

File: internal/compiler/generator/cpp.rs (hunk 1 above)
Comment (tronical) [reply internal/compiler/generator/cpp.rs:108]: Ahh, `TypeAlias`, yes. Good catch!

File: api/cpp/cmake/SlintMacro.cmake [hunk 5]
```diff
@@ -48,8 +57,15 @@ function(SLINT_TARGET_SOURCES target)
         set(scale_factor_target_prop "$<TARGET_PROPERTY:${target},SLINT_SCALE_FACTOR>")
         set(scale_factor_arg "$<IF:$<STREQUAL:${scale_factor_target_prop},>,,--scale-factor=${scale_factor_target_prop}>")
 
+        if (compilation_units GREATER 0)
+            foreach(cpp_num RANGE 1 ${compilation_units})                
+                list(APPEND cpp_files "${CMAKE_CURRENT_BINARY_DIR}/slint_generated_${_SLINT_BASE_NAME}_${cpp_num}.cpp")
```
Comment (ogoffart) [inline api/cpp/cmake/SlintMacro.cmake:62]: idea: name it `foobar.slint.cpp`?
But your slint_generated_ is not bad either.

File: tools/compiler/main.rs [hunk 6]
```diff
@@ -89,8 +93,29 @@ fn main() -> std::io::Result<()> {
...
+            generator::OutputFormat::Cpp(ref mut config) => {
+                config.cpp_files = args.cpp_files;
+
+                if args.output == std::path::Path::new("-") {
+                    eprintln!("--cpp-file can only be used together with -o");
+                    std::process::exit(1);
+                }
+
+                config.header_include =
+                    args.output.file_name().unwrap().to_string_lossy().to_string();
```
[the remark this answers is not in the document]
Comment (tronical) [reply tools/compiler/main.rs]: We do make sure of the include path but only in the cmake bits. It's true, we could use an absolute path. I'm a bit hesitant about that always - but given that the generated ninja files have absolute paths all over the place... I think you're right, that's the safer option. I'll do that.

File: api/cpp/cmake/SlintMacro.cmake (hunk 5 above)
Comment (hunger) [reply api/cpp/cmake/SlintMacro.cmake:62]: Iprefer `slint_generated_` as a prefix. It mages it very clear that it is generated and was done by slint.

**Comment by hunger** [comment]
Considering that we ask for C++20 anyway: Can we generate a C++ module?

From what I read using random modules is still a bit off, but modularizing your own code seems to be doable nowadays. But that is all we need.

**Comment by tronical** [comment]
Requiring C++ module support at this point would sadly mean excluding various compilers (including for example the xtensa toolchain from Espressif), so that's not an option at this point. We could consider implementing C++ module support as an opt-in, but I would not prioritise that work at all right now - especially given that it adds additional maintenance cost. The compiler generated code would have to be able to generate regular header files as well as a module header.

That said, what would be the benefit of generating a C++ module here? Is it about separating public generated API from internal?

**Comment by hunger** [comment]
Yes, this would definitely need an extra compiler flag or something to turn on. We can not just switch to modules and forget about existing users using the traditional approach. I was just wondering :-)

**Pull request merged**